### PR TITLE
Null annotate VBCSCompiler

### PR DESCRIPTION
**Pull request by jaredpar**
In attempting to clean up the logging code today I got stuck on a number
of cases where I was uncertain about `null` values. Decided to take the
time to finish `null` annotating this code base.

### CODE REVIEW

File: src/Compilers/Core/CommandLine/BuildProtocol.cs [hunk 1]
```diff
@@ -72,9 +72,9 @@ internal class BuildRequest
 
         public static BuildRequest Create(RequestLanguage language,
                                           IList<string> args,
-                                          string? workingDirectory = null,
-                                          string? tempDirectory = null,
-                                          string? compilerHash = null,
+                                          string workingDirectory,
+                                          string tempDirectory,
+                                          string compilerHash,
```
Comment (jaredpar) [inline src/Compilers/Core/CommandLine/BuildProtocol.cs:77]: These had been previously annotated as allowing `null`  but upon further inspection it's clear that `null` is not allowed as a value here. The client never provides `null` and the server actively rejects requests that provide `null` here.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs [hunk 2]
```diff
@@ -106,7 +106,7 @@ public sealed partial class AnalyzerConfig
         /// Parses an editor config file text located at the given path. No parsing
         /// errors are reported. If any line contains a parse error, it is dropped.
         /// </summary>
-        public static AnalyzerConfig Parse(string text, string pathToFile)
+        public static AnalyzerConfig Parse(string text, string? pathToFile)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs:109]: In annotating `CommonCompiler` and looking at other call sites it's clear that `null` is an allowed value here.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs [hunk 3]
```diff
@@ -115,9 +115,9 @@ public static AnalyzerConfig Parse(string text, string pathToFile)
         /// Parses an editor config file text located at the given path. No parsing
         /// errors are reported. If any line contains a parse error, it is dropped.
         /// </summary>
-        public static AnalyzerConfig Parse(SourceText text, string pathToFile)
+        public static AnalyzerConfig Parse(SourceText text, string? pathToFile)
         {
-            if (!Path.IsPathRooted(pathToFile) || string.IsNullOrEmpty(Path.GetFileName(pathToFile)))
+            if (pathToFile is null || !Path.IsPathRooted(pathToFile) || string.IsNullOrEmpty(Path.GetFileName(pathToFile)))
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs:120]: `IsPathRooted` already handles `null` but that that doesn't help the analysis below hence added an explicit `null` check.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 4]
```diff
@@ -39,15 +38,15 @@ namespace Microsoft.CodeAnalysis
...
         /// </summary>
-        internal string SdkDirectory { get; }
+        internal string? SdkDirectory { get; }
 
         /// <summary>
         /// The temporary directory a compilation should use instead of <see cref="Path.GetTempPath"/>.  The latter
         /// relies on global state individual compilations should ignore.
         /// </summary>
-        internal string TempDirectory { get; }
+        internal string? TempDirectory { get; }
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:47]: Considered calling this `CustomTempDirectory` as it's really used when we override `Path.GetTempPath`. I think it makes the intent clearer. Open to doing this but wanted to hear feedback from others.

Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:47]: Or maybe `ExplicitTempDirectory`

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 5]
```diff
@@ -416,11 +414,12 @@ internal EmbeddedText TryReadEmbeddedFileContent(string filePath, DiagnosticBag
             }
         }
 
-        private ImmutableArray<EmbeddedText> AcquireEmbeddedTexts(Compilation compilation, DiagnosticBag diagnostics)
+        private ImmutableArray<EmbeddedText?> AcquireEmbeddedTexts(Compilation compilation, DiagnosticBag diagnostics)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:417]: It seems really odd that we explicitly allow for `null` values in this collection but we clearly do (see comment at the bottom of this method). Had thoughts about undoing that but felt it was too much for this PR.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 6]
```diff
@@ -1062,12 +1065,12 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
 
             cancellationToken.ThrowIfCancellationRequested();
 
-            string outputName = GetOutputFileName(compilation, cancellationToken);
+            string outputName = GetOutputFileName(compilation, cancellationToken)!;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:1068]: This is the hardest to grok and I want to come back and refactor this later. For now just being honest about the `null` intent here.

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler PTAL

**Comment by 333fred** [comment]
@jaredpar test failures look legitimate.

**Comment by jaredpar** [comment]
Finally got a build of VS where the test explorer was working and was able to fix the test issues here. @dotnet/roslyn-compiler PTAL

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 7]
```diff
@@ -154,12 +150,13 @@ internal string GetCompilerVersion()
...
-            string assemblyVersion = GetInformationalVersionWithoutHash(type);
-            string hash = GetShortCommitHash(type);
+            string? assemblyVersion = GetInformationalVersionWithoutHash(type);
+            string? hash = GetShortCommitHash(type);
             return $"{assemblyVersion} ({hash})";
         }
 
-        internal static string ExtractShortCommitHash(string hash)
+        [return: NotNullIfNotNull("hash")]
+        internal static string? ExtractShortCommitHash(string hash)
```
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:159]: >string [](start = 55, length = 6)

`string?`?

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 8]
```diff
@@ -170,17 +167,19 @@ internal static string ExtractShortCommitHash(string hash)
...
             // We remove the section after the + (if any is present)
             return type.Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion.Split('+')[0];
         }
 
-        private static string GetShortCommitHash(Type type)
+        private static string? GetShortCommitHash(Type type)
         {
             var hash = type.Assembly.GetCustomAttribute<CommitHashAttribute>()?.Hash;
-            return ExtractShortCommitHash(hash);
+            return hash is object
```
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:180]: >hash is object [](start = 19, length = 14)

`ExtractShortCommitHash()` handles `null` value.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:417]: > It seems really odd that we explicitly allow for null values in this collection ...

Agreed, this seems odd. Especially since the collection is passed to other types (such as to `Compilation.CheckOptionsAndCreateModuleBuilder()`).

Could we just replace the `ImmutableArray.CreateBuilder<EmbeddedType?>(...)` with `ArrayBuilder.CreateInstance<EmbeddedText>()` and avoid adding to that builder if `text is null`?

---
In reply to: [502079305](https://github.com/dotnet/roslyn/pull/48460#discussion_r502079305) [](ancestors = 502079305)

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 9]
```diff
@@ -1566,12 +1570,12 @@ private static Stream OpenStream(CommonMessageProvider messageProvider, string p
             return null;
         }
 
-        private static string ResolveRelativePath(CommonMessageProvider messageProvider, string path, string baseDirectory, DiagnosticBag diagnostics)
+        private static string? ResolveRelativePath(CommonMessageProvider messageProvider, string? path, string? baseDirectory, DiagnosticBag diagnostics)
```
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:1573]: >string? [](start = 90, length = 7)

Is this change needed? It looks like the only caller is above which checks `path == null`.

File: src/Compilers/Server/VBCSCompilerTests/ServerUtil.cs [hunk 10]
```diff
@@ -23,6 +23,8 @@ namespace Microsoft.CodeAnalysis.CompilerServer.UnitTests
 {
     internal static class ProtocolUtil
     {
+        internal const string DefaultWorkingDirectory = @"c:\code";
```
Comment (cston) [inline src/Compilers/Server/VBCSCompilerTests/ServerUtil.cs:26]: >internal [](start = 8, length = 8)

It seems strange to see an invalid path used as a fallback value. Could we remove this constant and use `Path.GetTempPath()` (in the one use) below instead?

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 9 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:1573]: It's not strictly necessary. Was more trying to match the overall pattern of `path` and `baseDirectory` in the file. Will remove.

File: src/Compilers/Server/VBCSCompilerTests/ServerUtil.cs (hunk 10 above)
Comment (jaredpar) [reply src/Compilers/Server/VBCSCompilerTests/ServerUtil.cs]: Reworked this a bit to avoid the fake paths. In some ways this pattern would be ideal if I could guarantee that was an invalid directory but cannot so went with a `TempRoot` solution.

**Comment by jaredpar** [comment]
Responded to PR feedback.

**Review by cston — APPROVED** [review]
LGTM.

As we discussed, it would be good to fix https://github.com/dotnet/roslyn/pull/48460#discussion_r503469358, perhaps in a separate PR.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs:120]: It should. [https://github.com/dotnet/runtime/commit/25ea6c44cf3140cc9bfbaaff7f01ad2a27414f8e](https://github.com/dotnet/runtime/commit/25ea6c44cf3140cc9bfbaaff7f01ad2a27414f8e) was over a year ago.

---
In reply to: [502078334](https://github.com/dotnet/roslyn/pull/48460#discussion_r502078334) [](ancestors = 502078334)

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:47]: Eh. I think it's fine.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 11]
```diff
@@ -311,7 +307,7 @@ internal SourceText TryReadFileContent(CommandLineSourceFile file, IList<Diagnos
                 var directory = Path.GetDirectoryName(normalizedPath) ?? normalizedPath;
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:307]: It looks like `normalizedPath` is never null here, or `fileContent` would have been null. Consider asserting and then removing the check below.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:1066]: Is there a bug tracking?

---
In reply to: [502079520](https://github.com/dotnet/roslyn/pull/48460#discussion_r502079520) [](ancestors = 502079520)

File: src/Compilers/Server/VBCSCompiler/MetadataCache.cs [hunk 12]
```diff
@@ -48,7 +46,7 @@ private ImmutableArray<ModuleMetadata> GetAllModules(ModuleMetadata manifestModu
                     moduleBuilder.Add(manifestModule);
                 }
 
-                var module = CreateModuleMetadata(PathUtilities.CombineAbsoluteAndRelativePaths(assemblyDir, moduleName), prefetchEntireImage: false);
+                var module = CreateModuleMetadata(PathUtilities.CombineAbsoluteAndRelativePaths(assemblyDir, moduleName)!, prefetchEntireImage: false);
```
Comment (333fred) [inline src/Compilers/Server/VBCSCompiler/MetadataCache.cs:49]: Why does this need a suppression?

File: src/Compilers/Server/VBCSCompiler/MetadataCache.cs [hunk 13]
```diff
@@ -74,10 +72,11 @@ internal Metadata GetMetadata(string fullPath, MetadataReferenceProperties prope
             }
             else
             {
+                Debug.Assert(fileKey.HasValue);
                 var primaryModule = CreateModuleMetadata(fullPath, prefetchEntireImage: false);
 
                 // Get all the modules, and load them. Create an assembly metadata.
-                var allModules = GetAllModules(primaryModule, Path.GetDirectoryName(fullPath));
+                var allModules = GetAllModules(primaryModule, Path.GetDirectoryName(fullPath)!);
```
Comment (333fred) [inline src/Compilers/Server/VBCSCompiler/MetadataCache.cs:79]: >Path.GetDirectoryName(fullPath)! [](start = 62, length = 32)

Why does this need a suppression?

**Comment by 333fred** [comment]
Done review pass (commit 6)

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs:120]: Nearly a year ago but not in netcoreapp3.1

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 6 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:1066]: https://github.com/dotnet/roslyn/issues/48599

File: src/Compilers/Server/VBCSCompiler/MetadataCache.cs (hunk 12 above)
Comment (jaredpar) [reply src/Compilers/Server/VBCSCompiler/MetadataCache.cs:49]: The parameter requires `string` and there is no way to annotate `CombineAbsoluteAndRelativePath` to return a non-null value. Can't even use `NotNullIfNotNull` here because it will return `null` based on the type of path in the second parameter.

File: src/Compilers/Server/VBCSCompiler/MetadataCache.cs (hunk 13 above)
Comment (jaredpar) [reply src/Compilers/Server/VBCSCompiler/MetadataCache.cs:79]: Because `GetDirectoryName` returns `string?` (both now and in net5.0).

**Comment by jaredpar** [comment]
Responded to PR feedback

**Review by 333fred — APPROVED** [review]
LGTM (commit 7)

**Comment by jaredpar** [comment]
Failure is the inability to clean up listeners failure.
